pkg/client/cache/store.go

```diff
@@ -27,40 +28,61 @@ import (
 // as a local caching system, and an LRU store, which allows Reflector to work like a
 // queue of items yet to be processed.
 type Store interface {
-	Add(id string, obj interface{})
-	Update(id string, obj interface{})
-	Delete(id string)
+	Add(obj interface{})
+	Update(obj interface{})
+	Delete(obj interface{})
 	List() []interface{}
-	ContainedIDs() util.StringSet
-	Get(id string) (item interface{}, exists bool)
+	Get(obj interface{}) (item interface{}, exists bool)
 
 	// Replace will delete the contents of the store, using instead the
-	// given map. Store takes ownership of the map, you should not reference
+	// given list. Store takes ownership of the list, you should not reference
 	// it after calling this function.
-	Replace(idToObj map[string]interface{})
+	Replace([]interface{})
+}
+
+type KeyFunc func(obj interface{}) (string, error)
+
+func NameKeyFunc(obj interface{}) (string, error) {
```

Comment:
Is there any case where using NameKeyFunc as your keyfunc does the right thing and using NamespaceKeyFunc does the wrong thing?   If not, don't provide NameKeyFunc.